File: app/controllers/projects_controller.rb
class ProjectsController < ApplicationController
  before_filter :authenticate_user!

  # GET /projects/:id
  def show
    @project = Project.find(params[:id])

    respond_to do |format|
      format.json { render json: @project }
    end
  end

  # GET /projects
  def index
    @projects = Project.all
    
    respond_to do |format|
      format.json { render json: @projects }
    end
  end

  # POST /projects
  def create
    @project = Project.new(project_params)

    respond_to do |format|
      if @project.save
        format.json { render json: @project, status: :created }
      else
        format.json { render json: @project.errors, status: :unprocessable_entity }
      end
    end
  end

  # PUT /projects/:id
  def update
    @project = Project.find(params[:id])

    respond_to do |format|
      if @project.update(project_params)
        format.json { render json: nil, status: :ok }
      else
        format.json { render json: @project.errors, status: :unprocessable_entity}
      end
    end
  end

  # DELETE /projects/:id.json
  def destroy
    @project = Project.find(params[:id])
    @project.destroy

    respond_to do |format|
      format.json { render json: nil, status: :ok }
    end
  end

  def project_params
    params.require(:project).permit(:id, :title, :author, :start_date, :publish_date, :notes)
  end
end

File: app/assets/javascripts/controllers/projects_controller.js
App.ProjectsController = Em.ArrayController.extend({
  newProjectModalButtons: [
    Ember.Object.create({title: 'Create', clicked: 'createProject'}),
    Ember.Object.create({title: 'Cancel', clicked: 'cancel', dismiss: 'modal'})
  ],
  actions: {
    deleteProject: function(project) {
      project.destroyRecord();
    },
    showNewProject: function(project) {
      if (project) {
        this.set('editProject', project);
        this.set('title',     project.get('title'));
        this.set('author', project.get('author'));
        this.set('startDate',      project.get('startDate'));
        this.set('publishDate',  project.get('publishDate'));
        this.set('notes',    project.get('notes'));
      }
      return Bootstrap.ModalManager.show('newProject');
    },
    cancel: function() {
      this.set('title', '');
      this.set('author', '');
      this.set('startDate', '');
      this.set('publishDate', '');
      this.set('notes', '');
    },
    createProject: function() {
      newProject = this.get('editProject');
      this.set('errors', {}); // move validation into the controller
      if (validate(this, this.get('validations'))) {
        if (!newProject) { // if newProject is undefined
          newProject = this.store.createRecord('project', {
            title:       this.get('title'),
            author:      this.get('author'),
            startDate:   this.get('startDate'),
            publishDate: this.get('publishDate'),
            notes:       this.get('notes')
          });
        }
        // if defined, then editing an existing member so update all fields
        else { 
          newProject.set('title', this.get('title'));
          newProject.set('author', this.get('author'));
          newProject.set('startDate', this.get('startDate'));
          newProject.set('publishDate', this.get('publishDate'));
          newProject.set('notes', this.get('notes'));
        }

        var controller = this;
        newProject.save().then(function() {
          controller.send('cancel');
          return Bootstrap.NM.push('Succesfully added ' + controller.get('title') + '.', 'success');
        }, function() {
          return Bootstrap.NM.push('Failed to add ' + controller.get('title') + '.', 'danger');
        });

        return Bootstrap.ModalManager.close('newProject');
      }
    },
  },
  validations: {
    title: {
      regex: /^[A-Za-z0-9 ]{3,20}$/, 
      message: "Enter a title (min of 3 characters, max of 20)"
    },
    author: {
      regex: /.*/,
      message: "Enter an author for this project"
    },
    startDate: {
      regex: /\d\d\d\d-\d\d-\d\d/,
      message: "Enter a valid date"
    }
  }
});
